### PR TITLE
[Refactor] Enums - Psychrometrics.hh

### PR DESCRIPTION
**Pull request by jmythms**
Pull request overview
---------------------
1. Converts the constexpr ints in Psychrometrics.hh to enums. 
2. Converts ObjexxFCL arrays to std::arrays. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by amirroth** [comment]
I would like to throw a couple more things into this branch if I can:

- Replace the `cached_Tsat`, `cached_Psat`, etc. arrays themselves with `std::array`s.
- Replace function local `RoutineName`s with references to `PsyRoutineNames[]`.

File: src/EnergyPlus/PsychCacheData.hh [hunk 1]
```diff
@@ -121,40 +163,40 @@ struct PsychrometricCacheData : BaseGlobalStruct
...
-        cached_Psat.clear();
+        cached_Psat.fill(cached_psat_t());
 #endif
 #ifdef EP_cache_PsyTsatFnPb
-        cached_Tsat.clear();
+        cached_Tsat.fill(cached_tsat_h_pb());
 #endif
 #ifdef EP_cache_PsyTsatFnHPb
-        cached_Tsat_HPb.clear();
+        cached_Tsat_HPb.fill(cached_tsat_h_pb());
```
Comment (jmythms) [inline src/EnergyPlus/PsychCacheData.hh]: `std::array` does not have a `.clear()` method. So, I `.fill()`ed them with the default constructor call for those structs.

File: src/EnergyPlus/Psychrometrics.cc [hunk 2]
```diff
@@ -171,16 +119,16 @@ namespace Psychrometrics {
...
-        state.dataPsychCache->cached_Psat.allocate({0, psatcache_size});
+        state.dataPsychCache->cached_Psat.fill(cached_psat_t());
 #endif
 #ifdef EP_cache_PsyTsatFnPb
-        state.dataPsychCache->cached_Tsat.allocate({0, tsatcache_size});
+        state.dataPsychCache->cached_Tsat.fill(cached_tsat_h_pb());
 #endif
 #ifdef EP_cache_PsyTsatFnHPb
-        state.dataPsychCache->cached_Tsat_HPb.allocate({0, tsat_hbp_cache_size});
+        state.dataPsychCache->cached_Tsat_HPb.fill(cached_tsat_h_pb());
```
Comment (jmythms) [inline src/EnergyPlus/Psychrometrics.cc]: `.fill()`ed with the default constructor call for these structs.

File: src/EnergyPlus/Psychrometrics.cc [hunk 3]
```diff
@@ -449,7 +400,7 @@ namespace Psychrometrics {
 
         // FUNCTION PARAMETER DEFINITIONS:
         int constexpr itmax(100); // Maximum No of Iterations
-        static constexpr std::string_view RoutineName("PsyTwbFnTdbWPb");
+        static constexpr std::string_view RoutineName(PsyRoutineNames[static_cast<int>(PsychrometricFunction::RhFnTdbRhov)]); // PsyTwbFnTdbWPb
```
Comment (jmythms) [inline src/EnergyPlus/Psychrometrics.cc:403]: Keep the `static` keyword? I left `RoutineName`s as separate variables for readability.

File: src/EnergyPlus/Psychrometrics.hh [hunk 4]
```diff
@@ -845,10 +844,10 @@ namespace Psychrometrics {
         // values from PsyRhFnTdbWPb
 
         // FUNCTION PARAMETER DEFINITIONS:
-        static std::string const RoutineName("PsyRhFnTdbRhov");
+        static constexpr std::string_view RoutineName(PsyRoutineNames[static_cast<int>(PsychrometricFunction::RhFnTdbRhov)]); // PsyRhFnTdbRhov
```
Comment (jmythms) [inline src/EnergyPlus/Psychrometrics.hh]: Keep the `static` keyword? I left `RoutineName`s as separate variables for readability.

File: src/EnergyPlus/PsychCacheData.hh [hunk 5]
```diff
@@ -66,6 +89,25 @@ constexpr int NumPsychMonitors = 19; // Parameterization of Number of psychromet
...
+#endif
+#ifdef EP_cache_PsyTsatFnPb
+constexpr int tsatcache_size = 1024 * 1024;
+constexpr int tsatprecision_bits = 24;
+constexpr Int64 tsatcache_mask = tsatcache_size - 1;
+#endif
+#ifdef EP_cache_PsyTsatFnHPb
+constexpr int tsat_hbp_cache_size = 1024 * 1024;
+constexpr int tsat_hbp_precision_bits = 28;
+#endif
```
Comment (Myoldmopar) [inline src/EnergyPlus/PsychCacheData.hh:109]: `constexpr` FTW!

File: src/EnergyPlus/PsychCacheData.hh [hunk 6]
```diff
@@ -121,40 +163,28 @@ struct PsychrometricCacheData : BaseGlobalStruct
 {
 
 #ifdef EP_cache_PsyTwbFnTdbWPb
-    Array1D<cached_twb_t> cached_Twb; // DIMENSION(0:twbcache_size)
+    std::array<cached_twb_t, twbcache_size> cached_Twb;
```
Comment (Myoldmopar) [inline src/EnergyPlus/PsychCacheData.hh:166]: Yes, very very much better!

File: src/EnergyPlus/Psychrometrics.hh [hunk 7]
```diff
@@ -105,58 +105,57 @@ namespace Psychrometrics {
...
+        "PsyTsatFnHPb",
+        "PsyTsatFnPb",
+        "PsyRhFnTdbRhov",
+        "PsyRhFnTdbRhovLBnd0C",
+        "PsyTwbFnTdbWPb",
+        "PsyTwbFnTdbWPb",
+        "PsyWFnTdbTwbPb",
+        "PsyTsatFnPb",
+        "PsyTwbFnTdbWPb_cache",
+        "PsyPsatFnTemp_cache"}; // 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 |
```
Comment (Myoldmopar) [inline src/EnergyPlus/Psychrometrics.hh:128]: With this initialization, I wonder if the comment block at the end is really necessary.  I suggest either moving the comments up to each individual element, or delete the comment block if not necessary.

File: src/EnergyPlus/Psychrometrics.hh [hunk 8]
```diff
@@ -105,58 +105,57 @@ namespace Psychrometrics {
...
+        true,
+        true,
+        true,
+        true,
+        false,
+        false,
+        false,
+        false,
+        true,
+        true}; // PsyTdpFnTdbTwbPb     1 | PsyRhFnTdbWPb        2 | PsyTwbFnTdbWPb       3 | PsyVFnTdbWPb         4 |
```
Comment (Myoldmopar) [inline src/EnergyPlus/Psychrometrics.hh:153]: Same with this comment block.  On this one it is probably worth pulling each one up to its associated element for clarity.

File: src/EnergyPlus/Psychrometrics.hh [hunk 9]
```diff
@@ -513,7 +512,7 @@ namespace Psychrometrics {
         // ASHRAE handbook 1993 Fundamentals,
 
 #ifdef EP_psych_stats
-        ++state.dataPsychCache->NumTimesCalled(iPsyRhFnTdbRhovLBnd0C);
+        ++state.dataPsychCache->NumTimesCalled[static_cast<int>(PsychrometricFunction::RhFnTdbRhovLBnd0C)];
```
Comment (Myoldmopar) [inline src/EnergyPlus/Psychrometrics.hh:515]: I believe this is the pattern we are going for, so this is good.

**Review by Myoldmopar — APPROVED** [review]
This looks fine to me.  There are a couple tiny comment-only changes that don't necessarily need to hold up the PR.  If anyone else has any comments to add, and any commits are needed, then go ahead and clean those comments at the same time.  Otherwise I'll do final testing and see if we can get this merged directly.

**Comment by Myoldmopar** [comment]
I pulled develop in, and I want to wait until the GHA alternate build configurations run on this branch because some of the conflicts were instead of #ifdef scopes that I'm not building.  If it is happy, I believe everything is good to go here.

**Comment by Myoldmopar** [comment]
So far so good.  I'm going to go ahead and let Decent catch up, but I suspect I'll merge this either tonight or tomorrow morning.  Thanks @jmythms !

**Comment by Myoldmopar** [comment]
Well, unfortunately develop moved again.  Those diffs are unrelated.  I can run regressions locally on this branch to get a good set of results.

**Comment by Myoldmopar** [comment]
I just built and ran the full test suite on my machine and got zero diffs.  This is good to go.  Thanks @jmythms

**Comment by jmythms** [comment]
Thank you!

**Pull request merged**